### PR TITLE
chore: Refactor cell recovery code

### PR DESCRIPTION
**Pull request by kevaundray**
This PR modularizes what was previously being called a "shift on a polynomial" to the coset_fft method. 

Initially it was named "shift_fft_field" so that it is not confused with the cosets created by reverse_bit_ordering, but left it as this because this is the conventional name.

Changes made:

- Remove shift_polynomial_coeff
- Remove recover_shifted_data
- Remove recover_original_data
- Move `zero_poly_eval_brp ` under sanity check comment as its only used for sanity checking

### CODE REVIEW

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (kevaundray) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:23]: Theres a test_fft method, so blocking this until we add a similar method to test `coset_fft`

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (jtraglia) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:595]: There's a missing a word at the end ("evaluations missing" vs "evaluations are missing") but I think this could be phrased a little differently for better clarity.
```suggestion
    Recover the missing evaluations for the extended blob, given at least half of the evaluations.
```

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (jtraglia) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:567]: Just a nit, unrelated to your changes.
```suggestion
    Given the cells IDs that are missing from the data, compute the polynomial that vanishes at every point that
```

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (jtraglia) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:192]: I'm not opposed to it, but I don't think we typically use nested functions in the specs.

**Review by jtraglia — COMMENTED** [review]
I think this refactor is cleaner, thanks!

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (kevaundray) [reply specs/_features/eip7594/polynomial-commitments-sampling.md:23]: unblocked with [92d3fcb](https://github.com/ethereum/consensus-specs/pull/3781/commits/92d3fcbc4a4336e6837cc4cc45cffb5b62fee0d9)

**Review by asn-d6 — COMMENTED** [review]
Initial pass.

**Review by b-wagn — COMMENTED** [review]
I left some minor comments.
In general, I think this change improves the structure a lot. :+1:

File: tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py
Comment (jtraglia) [reply tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py:54]: Good catch. I tested this & it fails if `num_missing_cells` is set to `spec.CELLS_PER_EXT_BLOB`.

image

[This is a limitation in c-kzg too](https://github.com/ethereum/c-kzg-4844/blob/c7bc31206cd3dbff52076dfe20b20931d7ffbc47/src/c_kzg_4844.c#L2631). But we require [there are at least 50% of cells present](https://github.com/ethereum/c-kzg-4844/blob/c7bc31206cd3dbff52076dfe20b20931d7ffbc47/src/c_kzg_4844.c#L2886) before it's used.

I think we should add this restriction to `construct_vanishing_polynomial`'s docstring.

File: tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py
Comment (kevaundray) [reply tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py:46]: I think it would be good to do this in a separate PR since this is something that was present before this PR with test_fft

Comment (b-wagn) [reply tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py:46]: Yes, I agree. 
For such a PR it would be helpful if your PR defines the shift factor in some way that can then be accessed by these tests.

File: tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py
Comment (kevaundray) [reply tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py:54]: I don't think it should be a restriction -- will modify the method to handle that case

Comment (kevaundray) [reply tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py:54]: I added the code to handle that path here: https://github.com/ethereum/consensus-specs/pull/3781/commits/b0be78af79479498d34f0271ba638f6690af2801

Though I can see an argument to removing it -- on one hand, it makes the function work on all inputs, though we never hit that case that is added. Let me know what you think @jtraglia and @b-wagn

Comment (b-wagn) [reply tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py:54]: I think making the function work for all inputs makes sense. Otherwise, we would have to document and explain why it does not work for this one case.

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (asn-d6) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:186]: A bit of trailing whitespace here and along this function in general. Also in the docs of construct_vanishing_polynomial(). Maybe you should ask your editor to highlight them?

File: tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py
Comment (asn-d6) [reply tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py:54]: Personally, I prefer an assert that `len(missing_cell_ids) <= CELLS_PER_EXT_BLOB` and a remark about it on the func doc, compared to implementing something that will never ever be reached.

I understand it's nice to have the functions be complete, but at the same time this is a small protocol and this function is used once and hence we have a good understanding of its inputs.

Adding a block of spec code that will never be executed that needs to be implemented in three separate languages (including C), to me seems like a potential source of bugs for not much gain.

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (asn-d6) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:584]: Not a huge fan of this unreachable code. Not gonna die on this hill though. My comments on the thread below.

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (asn-d6) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:651]: H(X) seems to be some new notation introduced here. I tried to use Vitalik's post Q_1, Q_2, Q_3 notation, so that readers can follow. Perhaps it's worth mentioning them in the comments?

**Review by asn-d6 — CHANGES_REQUESTED** [review]
Thanks for this PR, @kevaundray ! I agree this is a better way at looking at (and implementing) the protocol.

I added some comments about the special-case vanishing poly code added in this ticket, that I would like to be discussed before merging this.

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (jtraglia) [reply specs/_features/eip7594/polynomial-commitments-sampling.md:651]: We also use this notation in ckzg:

image

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (kevaundray) [reply specs/_features/eip7594/polynomial-commitments-sampling.md:186]: Can you elaborate on the issue here? the linter is not showing up any issues

Comment (jtraglia) [reply specs/_features/eip7594/polynomial-commitments-sampling.md:186]: Just an extra space at the end. The linter won't complain about trailing whitespace.

image

image

Comment (jtraglia) [reply specs/_features/eip7594/polynomial-commitments-sampling.md:186]: Here's how you might auto-remove trailing whitespace in vscode:

* https://stackoverflow.com/a/30884298

File: tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py
Comment (kevaundray) [reply tests/core/pyspec/eth2spec/test/eip7594/unittests/polynomial_commitments/test_polynomial_commitments.py:46]: discussed offline with Ben to push this into the PR which will have the new test

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (kevaundray) [reply specs/_features/eip7594/polynomial-commitments-sampling.md]: Gracias

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (jtraglia) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:192]: If we want, we can delete the extra blank line here.

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (jtraglia) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:197]: I think the docstring should be the other style.
```suggestion
    def shift_vals(vals: Sequence[BLSFieldElement], factor: BLSFieldElement) -> Sequence[BLSFieldElement]:
        """
        Multiply each entry in `vals` by succeeding powers of `factor`
        i.e., [vals[0] * factor^0, vals[1] * factor^1, ..., vals[n] * factor^n]
        """
        shift = 1
```

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (jtraglia) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:581]: Something feels wrong the latter part of this sentence:

> and recovery only works then at most half of the cells are missing.

Maybe it should be:

> and recovery only works if half of the cells are available.

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (jtraglia) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:617]: Just a question. Zero is a valid evaluation though. What if all the evals for a cell are zero? Could that happen?

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (jtraglia) [inline specs/_features/eip7594/polynomial-commitments-sampling.md:701]: A bit hard to tell, but this can be joined onto a single line, right?

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (kevaundray) [reply specs/_features/eip7594/polynomial-commitments-sampling.md:581]: I can switch it around to say "at least" instead of "at most"

Comment (jtraglia) [reply specs/_features/eip7594/polynomial-commitments-sampling.md]: That would work. But my confusion is with the "works then at" part.

Comment (jtraglia) [reply specs/_features/eip7594/polynomial-commitments-sampling.md]: It might also be missing an "if" word. "only works then **if** at most".

File: specs/_features/eip7594/polynomial-commitments-sampling.md
Comment (kevaundray) [reply specs/_features/eip7594/polynomial-commitments-sampling.md:629]: Yep that could happen, if your blob is all zeroes and would be okay

**Review by jtraglia — APPROVED** [review]
LGTM! Thanks for making those changes 😄

**Review by asn-d6 — APPROVED** [review]
LGTM!

**Pull request merged**
